app/models/container_deployment.rb

```diff
@@ -6,7 +6,368 @@ class ContainerDeployment < ApplicationRecord
+             container_deployment_node.vm.hardware.ipaddresses.last unless container_deployment_node.tags.any? { |tag| tag.name.include?("deployment_master") }
+           elsif container_deployment_node.address
+             container_deployment_node.address unless container_deployment_node.tags.any? { |tag| tag.name.include?("deployment_master") }
+           end
+      ip_array << ip if ip
+    end
+    ip_array
+  end
+
+  def add_deployment_provider(options)
+    provider = ExtManagementSystem.new(
+      :type      => options[:provider_type],
+      :name      => options[:provider_name],
+      :port      => options[:provider_port],
+      :hostname  => options[:provider_hostname],
+      :ipaddress => options[:provider_ipaddress])
+    provider.save!
+    provider.update_authentication(:bearer => {:auth_key => options[:auth_key], :save => true})
+    valid_provider = provider.authentication_check.first
+    if valid_provider
+      self.deployed_ems = provider
+      save!
+    else
+      provider.destroy
+    end
+    valid_provider
+  end
+
+  def generate_ansible_inventory_for_subscription
+    template = <<EOS
+[OSEv3:children]
+nodes
+#{"masters" unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+
+[OSEv3:vars]
+ansible_ssh_user=#{ssh_auth.userid}
+deployment_type=openshift-enterprise
+rhsub_user=#{rhsm_auth.userid}
+rhsub_pass=#{rhsm_auth.password}
+rhsub_pool=#{rhsm_auth.rhsm_sku}
+ansible_sudo=true
+
+[nodes]
+#{(roles_addresses("node") - [roles_addresses("deployment_master")]).join("\n") unless roles_addresses("node").empty?}
+
+#{"[masters]" unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+#{(roles_addresses("master") - [roles_addresses("deployment_master")]).join("\n") unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+EOS
+    template
+  end
+
+  def generate_roles
+    result = {}
+    roles = container_deployment_nodes.collect(&:roles).flatten.uniq
+    result["master"] = {"osm_use_cockpit"                     => "false",
+                        "openshift_master_identity_providers" => [identity_ansible_config_format]} if roles.include? "master"
+    unless identity_provider_auth.first.htpassd_users.empty?
+      require 'htauth/md5'
+      salt = (0...8).map { (65 + rand(26)).chr }.join
+      md5 = HTAuth::Md5.new('salt' => salt)
+      htpass_hash = {}
+      users = []
+      identity_provider_auth.first.htpassd_users.each do |user|
+        users << JSON.parse(user.as_json.gsub(/\=\>/, ':'))
+      end
+      users.each do |user|
+        user["password"] = md5.encode(user["password"])
+        htpass_hash[(user['username']).to_s] = user["password"]
+      end
+      result["master"]["openshift_master_htpasswd_users"] = htpass_hash.to_json
+    end
+    result["node"] = {} if roles.include? "node"
+    result["storage"] = {} if roles.include? "storage"
+    result
+  end
+
+  def generate_ansible_yaml
```

Comment:
`generate_ansible_yaml` doesn't seem to be used outside of tests ... am I reading that right?  Do you need it for something else?